Documentation: examples of 'gold standard' slots as a reference guide for newcomers (#943)
GenomicsStandardsConsortium/mixs: src/docs/gold_standard_slot_examples.md

  - [Structured text (ontology term)](#structured-text-ontology-term)
  - [URL](#url) (or other link utilising a URL like structure)
  - [Mixed type (free text or ontology)](#mixed-type) (either free text, or from an ontology)

You can use these as templates on how to construct your own new slot to the highest quality level expected.

## Overview

Every MIxS LinkML slot should have at a minimum the following attributes:

- [`name`](https://linkml.io/linkml-model/latest/docs/name/): the term's structured comment name, written as the YAML key of the slot itself
- [`description`](https://linkml.io/linkml/schemas/metadata.html#providing-descriptions): the description of what the metadata term is for
- [`title`](https://linkml.io/linkml-model/latest/docs/title/): a short human readable 'title' for the slot
- [`examples`](https://linkml.io/linkml-model/latest/docs/examples/): example values demonstrating how the slot should be used
- [`slot_uri`](https://linkml.io/linkml-model/latest/docs/slot_uri/): a unique ID assigned by MIxS
  - This likely only gets assigned upon acceptance and merging by the core GSC MIxS team

>   - During development you can use a dummy value for this

Reviewer: This is the guidance that produced the placeholder IDs now sitting in `main`.

44 terms were merged carrying `MIXS:XXXXXXXXX`, identical on all 44, so they fail both the uniqueness rule and the format rule stated in section 8 of the term specifications. They arrived through https://github.com/GenomicsStandardsConsortium/mixs/pull/1230 and are tracked in https://github.com/GenomicsStandardsConsortium/mixs/issues/1304 and https://github.com/GenomicsStandardsConsortium/mixs/issues/1310.

As written, this page tells a newcomer to use a dummy value and never tells them it has to be replaced. Suggest adding that clause, something along the lines of: a dummy value is fine while drafting, but a MIxS ID assigned by the CIG must be in place before the pull request is merged.

Raising this as a comment rather than pushing a change, since this pull request is parked.

> # Gold Standard MIxS LinkML Slots Examples

Reviewer: Fixed.

## Introduction

This page lists a range of 'gold standard' examples of slots that are used in the MIxS LinkML schema.

> 

Reviewer: The scope note at the top now says this page shows how to write a new or revised term, and that existing terms may not match. https://github.com/GenomicsStandardsConsortium/mixs/pull/944 carries the normative rules.

> 

Reviewer: Both this page and the specification are now in the mkdocs nav under 'Writing terms'. The wider gap, that no hand-written doc in src/docs was linked, is https://github.com/GenomicsStandardsConsortium/mixs/issues/1294.

- Basic slots (i.e., match exactly to a LinkML `range` type):
  - [Integer](#integer) (numeric)
  - [Float](#float) (numeric)
  - [String](#string) (free text)
  - [Fixed list](#fixed-list) (enumeration)
  - [Boolean](#boolean)
- Complex (i.e., common more complex `range` types that are more specific to MIxS):
  - [Structured text (text only)](#structured-text-text-only) (text only)
  - [Structured text (with unit, measurement)](#structured-text-numeric-with-unit-measurement) (numeric with units)
  - [Structured text (ontology term)](#structured-text-ontology-term)
  - [URL](#url) (or other link utilising a URL like structure)
  - [Mixed type (free text or ontology)](#mixed-type) (either free text, or from an ontology)

You can use these as templates on how to construct your own new slot to the highest quality level expected.

## Overview

Every MIxS LinkML slot should have at a minimum the following attributes:

- [`description`](https://linkml.io/linkml/schemas/metadata.html#providing-descriptions): the description of what the metadata term is for
- [`title`](https://linkml.io/linkml-model/latest/docs/title/): a short human readable 'title' for the slot

> - [`examples`](https://linkml.io/linkml-model/latest/docs/examples/): examples values demonstrating how the slot should be used

Reviewer: Fixed.

> - [`in_subset`](https://linkml.io/linkml-model/latest/docs/in_subset/): the section of the schema that the slot belongs based on a [fix list of MIxS categories](https://github.com/GenomicsStandardsConsortium/mixs/blob/609b0f567486f64cb7061246588d8006f87fa138/src/mixs/schema/mixs.yaml#L21-L26)

Reviewer 2: I'm not sure we would want a contributor to determine the subset (categories). I also believe @only1chunts is working on some changes to these. So may be best to take this out, and make a issue to add this when it's cleared up.

Reviewer 2: also, just saw the note! :) 
So we could keep it

Author: Yes, I'm basing the current document on what we _currently_ have in the schema, this will of course evolve :)

Reviewer: The 'may be replaced in the near future' warning is removed. What a subset means is open in https://github.com/GenomicsStandardsConsortium/mixs/issues/931.

> - [`in_subset`](https://linkml.io/linkml-model/latest/docs/in_subset/): the section of the schema that the slot belongs based on a [fix list of MIxS categories](https://github.com/GenomicsStandardsConsortium/mixs/blob/609b0f567486f64cb7061246588d8006f87fa138/src/mixs/schema/mixs.yaml#L21-L26)

Reviewer: are we sure that we're going to continue to use subsets?

should be contrasted with any other hierarchy or grouping mechanisms we might want to use, like `is_a` or `slot_group`.

if we do use `in_subset` or `slot_group`, who can decide what belongs with what? and who gets to define the subset or group? failing to define a subset or grouping slot won't cause a build error (the way a dangling `is_as` would) but they are very bad style. we should check if the linter catches cases like those

Author: See the comment above to Montana/and mywarning below that it may be overhauled.

My assumption is that when someone makes a proposal, they can suggest which 'group' the term will belong to, and then will be confirmed  by CIG etc.

> - [`in_subset`](https://linkml.io/linkml-model/latest/docs/in_subset/): the section of the schema that the slot belongs based on a [fix list of MIxS categories](https://github.com/GenomicsStandardsConsortium/mixs/blob/609b0f567486f64cb7061246588d8006f87fa138/src/mixs/schema/mixs.yaml#L21-L26)

Reviewer: Fixed.

  - Note: this system may be replaced in the near future!

> - [`keywords`](https://linkml.io/linkml-model/latest/docs/keywords/): useful keywords to allow grouping of related slots together

Reviewer 2: ```suggestion
- [`keywords`](https://linkml.io/linkml-model/latest/docs/keywords/): useful keywords to allow grouping of related slots together
  - Note: Keywords should be applied from the existing and available keywords. Review and approval is required to add new keywords.
```

Reviewer 2: @turbomam to add issue to add a list of keywords

> - [`keywords`](https://linkml.io/linkml-model/latest/docs/keywords/): useful keywords to allow grouping of related slots together

Reviewer: When I introduced the keywords in 6.2, I didn't have any tools or policies for adding new keywords. And I still don't! So new terms could be added with no keywords, or with keywords that are similar to but different from the ones I used. Nothing is checking that yet. And we don't really have any easy tooling for retrieving all terms that are tagged with a keyword. That can be done with `SchemaView` loops, or by dumping MIxS to a tabular representation, but I don't think we have all agreed on the best way of doing that either.

Reviewer: Keywords are removed from this page. MIxS is retiring the assertions: https://github.com/GenomicsStandardsConsortium/mixs/issues/1283

- [`slot_uri`](https://linkml.io/linkml-model/latest/docs/slot_uri/): a unique ID assigned by MIxS

>   - This likely only gets assigned upon acceptance and merging by the core GSC MIxS team

Reviewer 2: ```suggestion
  - This ID will be assigned upon acceptance and merging by the core GSC MIxS team.
```

Reviewer 2: Note, we're not actually sure how the IDs are minted

Reviewer: I should make a tool that reports the highest id so far. I wonder if @sujaypatil96 could help me expose that on the web pages.

Author: I guess leaving it to the ultimate repsonsbility of the core MIxS team is then fine (they can always delegate)

Reviewer: IDs are assigned just before release, after a feature freeze. Answered on https://github.com/GenomicsStandardsConsortium/mixs/pull/944 by Lynn at GSC25.

  - During development you can use a dummy value for this

> - [`range`](https://linkml.io/linkml/schemas/slots.html#ranges): specifying the type of 'object' that put in the slot (numeric, string, etc.)

Reviewer: Fixed.

> 

Reviewer: The scope note at the top now says this page shows how to write a new or revised term, and that existing terms may not match. https://github.com/GenomicsStandardsConsortium/mixs/pull/944 carries the normative rules.

And for some slots, the following attributes are also recommended:

> - If a term should be either mandatory or optional

Reviewer 2: I would say this attribute is required for all slots. But the value may vary

Reviewer: Which attribute is required for all terms is settled in https://github.com/GenomicsStandardsConsortium/mixs/pull/944 section 4.1, and applies to new terms.

  - [`recommended`](https://linkml.io/linkml/schemas/slots.html#recommended): a boolean value indicating if the slot is recommended be filled

>   - [`required`](https://linkml.io/linkml/schemas/slots.html#required): a boolean value indicating if the slot is mandatory to be filled

Reviewer: I assume that we would never add any new terms with both recommended and required set. There are a few cases of that now that should be cleaned up. It is also possible that a term could be considered recommended in a Checklist and required in an Extension, so it would appear as both in a dynamically constructed combination class.

Reviewer 2: I think that's what the "either mandatory or optional" part it.
That said, I think we should review the terms that are present in checklists and extensions... I assume that shouldn't happen and if it does that seems confusing

Reviewer: Also, in terms of style, I don't think we should ever assert `false`. Any Boolean metaslot that isn't asserted is false by definition.

Author: Is the conclusion that what is added here is to just add 'recommended: true` if it is recommended etc., and just document you should never specify `recommended: false`?

Reviewer: Tracked in https://github.com/GenomicsStandardsConsortium/mixs/issues/1296.

- If a term should be allowed multiple entries

  - [`multivalued`](https://linkml.io/linkml/schemas/slots.html#multivalued)

>     - Essentially indicates the contents of the slot can be a list, and each element is evaluated independently against the remaining slot attributes

Reviewer: We need a lot more documentation about how multivalued values are represented in the various file formats/serializations that LinkML supports. In YAML or JSON the values would truly be broken out into a list

name: France
cities:
  - Paris
  - Lyon
  - Nice
  
  In CSV or TSV that is faithful to the current LinkML specifications, one would see

name | ciites
--- | ---
France | [Paris \| Lyon \| Nice ]

Whereas one might expect (or MIxS might imply) that the representation should/can omit the brackets

name | ciites
--- | ---
France | Paris \| Lyon \| Nice 

or even use a different delimiter

name | ciites
--- | ---
France | Paris ; Lyon ; Nice 

Much of that is under consideration in LinkML, but that kind of flexibility isn't supported yet. This can be seen first hand by creating some data that passes `linkml validate`, and then converting it with `linkml convert`, a process that I recommend to all MIxS contributors

Reviewer 2: Proposal, any multivalued slot must have 2 unique examples to show how to capture the information.
Note to keep this guidance focused on that individual that is proposing a new slot to GSC. 
This information is important for the individual that is using the standard and entering data. 
Maybe a separate section or issue or example file we can link to that would describe how to enter multivalued data. So that when someone is proposing a term, they can see some examples of structure & build their slot according to the need.

Author: @mslarae13 done here: https://github.com/GenomicsStandardsConsortium/mixs/pull/944/commits/a7239d26a17b46a948a178d733b8a8e4bdf9ccdc

Indeed, LinkML converts 

```yaml
human_gut_data:
  - samp_name: sample1
    project_name: project1
    special_diet:
      - low carb
      - vegetarian
  - samp_name: sample2
    project_name: project1
    special_diet:
      - low carb
      - reduced calorie

```

to 

```tsv
samp_name	project_name	special_diet
sample1	project1	[low carb|vegetarian]
sample2	project1	[low carb|reduced calorie]
```

Reviewer: Done in https://github.com/GenomicsStandardsConsortium/mixs/pull/944, section 7.3: a multivalued term must have at least two examples.

- If a string or mix-`range`-based term has a particular format it should follow:

  - [structured_pattern](https://linkml.io/linkml/schemas/constraints.html#structured-patterns): a particular regex-like pattern that includes pre-defined components that describe how each component should be formatted

>     - In the MIxS LinkML schema, these preset formats can be seen under the [`settings`](https://github.com/GenomicsStandardsConsortium/mixs/blob/609b0f567486f64cb7061246588d8006f87fa138/src/mixs/schema/mixs.yaml#L21852-L21887) section of the schema

Reviewer: Hopefully our structured_patterns won't look regex-like! They could be described as using a domain-specific language, or you could say that they compose the settings like macros?

Author: Maybe I'm too young (or skipped my excel days), but I've never used/seen a macro :sweat_smile: but can update teh text accordingly :+1:

Reviewer: Wording point taken: structured patterns compose named settings rather than being raw regex. https://github.com/GenomicsStandardsConsortium/mixs/pull/944 section 10.2 describes them that way.

  - [pattern](https://linkml.io/linkml/schemas/constraints.html#patterns): a regex pattern that the string will be compared against for validation
    - This is a more generalised version of `structured_pattern`, and can be used for any string-based slot
    - It is typically used for 'one-off' formats that would not be reused in other metadata schemas

## Examples

### Integer

An example of a close-to gold standard _integer_ slot is [`lib_reads_seqd`](https://genomicsstandardsconsortium.github.io/mixs/0000040/):

```yaml
lib_reads_seqd:
  description: Total number of clones sequenced from the library
  title: library reads sequenced
  examples:
    - value: "20"
  in_subset:
    - sequencing
  keywords:
    - library
  slot_uri: MIXS:0000040
  range: integer
```

This example slot allows a single bit of information in the form of a singular integer value.

<!--

> JFY comment: I don't like this so much as:

Reviewer 2: @jfy133 are you saying "let's pick a different slot for the integer example"?

Reviewer: Still open, and small: the integer example could use a better slot. Not a blocker. Worth a follow-up if anyone has a preferred candidate.

- the description is very minimal, the example value is in quotes
- single example (I like a few, even for very simple terms)
- only one keyword
- no recommended or required

> -->

Reviewer: Regarding quotes: the range of examples.value in LinkML is string, even if the range of the slot being illustrated is numeric or something else. In this case, the example could probably be written with without the quotation marks, but YAML has some idiosyncrasies regarding  automatic conversion of unquoted things like dates and Booleans, so I may have added the quotation marks as belt-and-suspenders insurance. Or maybe the conversion script did it.

Reviewer: The range of examples.value is string in LinkML regardless of the term's own range, so quoting is a YAML question rather than a typing one. Not a blocker here.

### Float

> 

Reviewer: Removed.

An example of a close-to gold standard _float_ slot is [`ph`](https://genomicsstandardsconsortium.github.io/mixs/0001001/):

```yaml
ph:
  description:
    Ph measurement of the sample, or liquid portion of sample, or aqueous
    phase of the fluid
  title: pH
  examples:
    - value: "7.2"
  keywords:
    - ph
  slot_uri: MIXS:0001001
  range: float
```

> This example slot allows a single bit of information in the form of a float value, i .e. allows decimal numbers.

Reviewer: Fixed.

  title: experimental factor
  examples:
    - value: time series design [EFO:0001779]
  in_subset:
    - investigation
  keywords:
    - experimental
    - factor
  string_serialization: "{termLabel} [{termID}]|{text}"
  slot_uri: MIXS:0000008
  multivalued: true
  range: string
  pattern: ^\S+.*\S+ \[[a-zA-Z]{2,}:\d+\]$
```

> [!WARNING]
> This slot example includes a `string_serialization:` attribute, which is discouraged as standard practice for a basic `pattern:` slot!

<!--
JFY comment: I don't like this so much as:

- single example
- Using a string_serialisation...
-->

### Structured text (numeric with unit, measurement)

An example of a close-to gold standard _Structured text (numeric with units)_ slot is [`size_frac`](https://genomicsstandardsconsortium.github.io/mixs/0000017/):

```yaml
size_frac:
  annotations:
    Expected_value: filter size value range
  description: Filtering pore size used in sample preparation
  title: size fraction selected
  examples:
    - value: 0-0.22 micrometer
  in_subset:
    - nucleic acid sequence source
  keywords:
    - fraction
    - size
  string_serialization: "{float}-{float} {unit}"
  slot_uri: MIXS:0000017

> ```

Reviewer: The scope note at the top now says this page shows how to write a new or revised term, and that existing terms may not match. https://github.com/GenomicsStandardsConsortium/mixs/pull/944 carries the normative rules.

-->

### Structured text (ontology term)

An example of a close-to gold standard _Structured text (ontology term)_ slot is [`env_medium`](https://genomicsstandardsconsortium.github.io/mixs/0000014/)

```yaml
env_medium:
  description: "Report the environmental material(s) immediately surrounding the
    sample or specimen at the time of sampling. We recommend using subclasses of
    'environmental material' (http://purl.obolibrary.org/obo/ENVO_00010483). EnvO
    documentation about how to use the field: https://github.com/EnvironmentOntology/envo/wiki/Using-ENVO-with-MIxS
    . Terms from other OBO ontologies are permissible as long as they reference
    mass/volume nouns (e.g. air, water, blood) and not discrete, countable entities
    (e.g. a tree, a leaf, a table top)"
  title: environmental medium
  examples:
    - value: bluegrass field soil [ENVO:00005789]
  in_subset:
    - environment
  keywords:
    - environmental
  slot_uri: MIXS:0000014
  range: string
  required: true
  pattern: ^([^\s-]{1,2}|[^\s-]+.+[^\s-]+) \[[a-zA-Z]{2,}:[a-zA-Z0-9]\d+\]$
  structured_pattern:
    syntax: ^{termLabel} \[{termID}\]$
    interpolated: true
    partial_match: true
```

> This example shows a specific type of structured text slot that is commonly used across in MIxS, where the value refers to a specific ontology term.

Reviewer: Fixed.

This particular pattern has the pattern of an ontology term label (a string) plus a unique identifier (a string) in square brackets.

> Implementers of the schema will often use such slots to perform API lookups to the ontology term against the ontology database, thus the pattern structure is critical when defining an ontology term-based metadata slot

Reviewer: Fixed.

<!--
JFY comment: I don't like this so much as:

- Why pattern AND structured_pattern?
-->

> ### URL

Reviewer 2: The structured pattern for this, and I think most slots that allow for URLs is "Structured pattern: ^{PMID}|{DOI}|{URL}$"

So an ID/DOI is also accepted... I wonder if we can refer to this as something else.. source? reference? IDK! But let's expand beyond URL and pick a slot with a URL, DOI, or PMID example.

Author: I don't follow @mslarae13 , sorry!

Do you mean a slot example with multiple examples? Or you mean update the header/title?

Reviewer: Still open: whether the URL example should show that a PMID or DOI is accepted too. Not a blocker for merging this page.

An example of a close-to gold standard _URL_ slot is [`isol_growth_condt`](https://genomicsstandardsconsortium.github.io/mixs/0000003/):

```yaml
isol_growth_condt:
  description:
    Publication reference in the form of pubmed ID (pmid), digital object
    identifier (doi) or url for isolation and growth condition specifications of
    the organism/material
  title: isolation and growth condition
  examples:
    - value: doi:10.1016/j.syapm.2018.01.009
  in_subset:
    - nucleic acid sequence source
  keywords:
    - condition
    - growth
    - isolation
  slot_uri: MIXS:0000003
  range: string
  pattern: ^^PMID:\d+$|^doi:10.\d{2,9}/.*$|^https?:\/\/(?:www\.)?[-a-zA-Z0-9@:%._\+~#=]{1,256}\.[a-zA-Z0-9()]{1,6}\b(?:[-a-zA-Z0-9()@:%_\+.~#?&\/=]*)$$
  structured_pattern:
    syntax: ^{PMID}|{DOI}|{URL}$
    interpolated: true
    partial_match: true
```

This example shows another common type of structured pattern for terms that refer to a URL-like or web address.

> The pattern in the example, allow a PubMed ID (`PMID:`), a digital object identifier (`doi:`), or a URL (uniform resource locator, `https:`), each with a specific prefix indicating the type of link address.

Reviewer: Fixed.
